Source/Scene/ShadowMap.js

```diff
@@ -307,110 +455,292 @@ define([
+                Matrix4.multiplyByVector(cascadeMatrix, corner, corner);
+                Cartesian3.divideByScalar(corner, corner.w, corner); // Handle the perspective divide
+                Cartesian3.minimumByComponent(corner, min, min);
+                Cartesian3.maximumByComponent(corner, max, max);
+            }
+
+            // Limit min to 0.0, sometimes precision errors cause it to go slightly negative.
+            min.x = Math.max(min.x, 0.0);
+            min.y = Math.max(min.y, 0.0);
+
+            // Always start cascade frustum at the top of the light frustum to capture objects in the light's path
+            min.z = 0.0;
+
+            var cascadeCamera = shadowMap._cascadeCameras[j];
+            cascadeCamera.clone(shadowMapCamera); // PERFORMANCE_IDEA : could do a shallow clone for all properties except the frustum
+            cascadeCamera.frustum.left = left + min.x * (right - left);
+            cascadeCamera.frustum.right = left + max.x * (right - left);
+            cascadeCamera.frustum.bottom = bottom + min.y * (top - bottom);
+            cascadeCamera.frustum.top = bottom + max.y * (top - bottom);
+            cascadeCamera.frustum.near = near + min.z * (far - near);
+            cascadeCamera.frustum.far = near + max.z * (far - near);
+
+            // Scale and offset are used to transform shadowPosition to the proper cascade in the shader
+            var cascadeScale = shadowMap._cascadeScales[j];
+            cascadeScale.x = 1.0 / (max.x - min.x);
+            cascadeScale.y = 1.0 / (max.y - min.y);
+            cascadeScale.z = 1.0 / (max.z - min.z);
+
+            var cascadeOffset = shadowMap._cascadeOffsets[j];
+            cascadeOffset.x = -min.x;
+            cascadeOffset.y = -min.y;
+            cascadeOffset.z = -min.z;
+        }
+    }
+
+    var scratchLightView = new Matrix4();
+    var scratchRight = new Cartesian3();
+    var scratchUp = new Cartesian3();
+    var scratchMin = new Cartesian3();
+    var scratchMax = new Cartesian3();
+    var scratchTranslation = new Cartesian3();
+
+    function fitShadowMapToScene(shadowMap) {
+        var shadowMapCamera = shadowMap._shadowMapCamera;
+        var sceneCamera = shadowMap._sceneCamera;
+
+        // 1. First find a tight bounding box in light space that contains the entire camera frustum.
+        var viewProjection = Matrix4.multiply(sceneCamera.frustum.projectionMatrix, sceneCamera.viewMatrix, scratchMatrix);
+        var inverseViewProjection = Matrix4.inverse(viewProjection, scratchMatrix);
+
+        // Start to construct the light view matrix. Set translation later once the bounding box is found.
+        var lightDir = shadowMapCamera.directionWC;
+        var lightUp = sceneCamera.directionWC; // Align shadows to the camera view.
+        var lightRight = Cartesian3.cross(lightDir, lightUp, scratchRight);
+        lightUp = Cartesian3.cross(lightRight, lightDir, scratchUp); // Recalculate up now that right is derived
```

Comment:
This may never be a problem, but keep in mind that `lightRight` and `lightUp` may not be unit length here due to floating-point error.